fdbserver/DataDistribution.actor.cpp

```diff
@@ -2343,93 +2364,131 @@ ACTOR Future<Void> scheduleAuditOnRange(Reference<DataDistributor> self,
+									continue;
+								}
+								req.targetServers.push_back(it->second[i].id());
+							}
+						} else if (audit->coreState.getType() == AuditType::ValidateLocationMetadata) {
+							auto it = rangeLocations[i].servers.begin(); // always do in primary DC
+							const int idx = deterministicRandom()->randomInt(0, it->second.size());
+							targetServer = it->second[idx];
+						} else {
+							UNREACHABLE();
+						}
+						// Set doAuditOnStorageServer
+						if (self->remainingBudgetForAuditTasks.contains(auditType)) {
+							ASSERT(self->remainingBudgetForAuditTasks[auditType].get() >= 0);
+							while (self->remainingBudgetForAuditTasks[auditType].get() == 0) {
+								wait(self->remainingBudgetForAuditTasks[auditType].onChange());
+								ASSERT(self->remainingBudgetForAuditTasks[auditType].get() >= 0);
+							}
+							self->remainingBudgetForAuditTasks[auditType].set(
+							    self->remainingBudgetForAuditTasks[auditType].get() - 1);
+							ASSERT(self->remainingBudgetForAuditTasks[auditType].get() >= 0);
+						} else {
+							self->remainingBudgetForAuditTasks[auditType].set(
+							    SERVER_KNOBS->CONCURRENT_AUDIT_TASK_COUNT_MAX - 1);
+						}
+						issueDoAuditCount++;
+						doAuditTasks.add(doAuditOnStorageServer(self, audit, targetServer, req));
 					}
-					self->remainingBudgetForAuditTasks[auditType].set(
-					    self->remainingBudgetForAuditTasks[auditType].get() - 1);
-					ASSERT(self->remainingBudgetForAuditTasks[auditType].get() >= 0);
-				} else {
-					self->remainingBudgetForAuditTasks[auditType].set(SERVER_KNOBS->CONCURRENT_AUDIT_TASK_COUNT_MAX -
-					                                                  1);
 				}
-				audit->actors.add(doAuditOnStorageServer(self, audit, targetServer, req));
+
 				// Proceed to the next range if getSourceServerInterfacesForRange is partially read
 				begin = rangeLocations[i].range.end;
 				wait(delay(0.1));
 			}
 		}
+
+		wait(doAuditTasks.getResult());
 		TraceEvent(SevDebug, "DDScheduleAuditOnRangeEnd", self->ddId)
 		    .detail("Reason", "End")
 		    .detail("AuditID", audit->coreState.id)
 		    .detail("Range", range)
 		    .detail("AuditType", auditType)
-		    .detail("IssuedDoAuditCount", audit->issueDoAuditCount);
+		    .detail("IssuedDoAuditCountInThisSchedule", issueDoAuditCount);
 
 	} catch (Error& e) {
-		TraceEvent(SevWarn, "DDScheduleAuditOnRangeError", self->ddId)
+		TraceEvent(SevInfo, "DDScheduleAuditOnRangeError", self->ddId)
 		    .errorUnsuppressed(e)
 		    .detail("AuditID", audit->coreState.id)
 		    .detail("Range", range)
-		    .detail("AuditType", auditType);
-		audit->anyChildAuditFailed = true;
+		    .detail("AuditType", auditType)
+		    .detail("IssuedDoAuditCount", issueDoAuditCount);
+		if (e.code() == error_code_actor_cancelled) {
+			throw e;
+		} else if (audit->retryCount < SERVER_KNOBS->AUDIT_RETRY_COUNT_MAX) {
+			audit->retryCount++;
+			audit->actors.add(scheduleAuditOnRange(self, audit, range));
```

Comment:
Is it possible to call scheduleAuditOnRange() within doAuditOnStorageServer() when it fails, since that is the minimum range we know to retry. If we retry here with range passed to scheduleAuditOnRange, we could restart multiple doAuditOnStorageServer?

Reply:
Yes.